### PR TITLE
DAOS-1742 md: rdb/raft regression test update

### PR DESCRIPTION
**Pull request by kccain**
Refactoring and enablement for multi-replica execution.
New multi-replica unit tests added.

Signed-off-by: Kenneth Cain <kenneth.c.cain@intel.com>

### CODE REVIEW

File: src/rdb/tests/rpc.h [hunk 1]
```diff
@@ -74,12 +113,75 @@ CRT_RPC_DECLARE(rdbt_init, DAOS_ISEQ_RDBT_INIT_OP, DAOS_OSEQ_RDBT_INIT_OP)
...
+	((struct rsvc_hint)	(tto_hint)		CRT_VAR) \
+	((uint64_t)		(tto_val)		CRT_VAR) \
 	((int32_t)		(tto_rc)		CRT_VAR)
 
 CRT_RPC_DECLARE(rdbt_test, DAOS_ISEQ_RDBT_TEST_OP, DAOS_OSEQ_RDBT_TEST_OP)
 
+#define DAOS_ISEQ_RDBT_MEMBERSHIP /* input fields */		 \
+	((d_rank_list_t)	(rtmi_ranks)		CRT_PTR)
+
+#define DAOS_OSEQ_RDBT_MEMBERSHIP /* output fields */		 \
```
[the remark this answers is not in the document]
Comment (kccain) [reply src/rdb/tests/rpc.h:158]: False positive for CaRT RPC definition related macros

File: src/rdb/tests/rpc.h [hunk 2]
```diff
@@ -74,12 +113,75 @@ CRT_RPC_DECLARE(rdbt_init, DAOS_ISEQ_RDBT_INIT_OP, DAOS_OSEQ_RDBT_INIT_OP)
 
 CRT_RPC_DECLARE(rdbt_fini, DAOS_ISEQ_RDBT_FINI_OP, DAOS_OSEQ_RDBT_FINI_OP)
 
+#define DAOS_ISEQ_RDBT_PING_OP	/* input fields (none) */
+
+#define DAOS_OSEQ_RDBT_PING_OP	/* output fields */		 \
```
[the remark this answers is not in the document]
Comment (kccain) [reply src/rdb/tests/rpc.h:118]: False positive for CaRT RPC definition related macros

File: src/rdb/tests/rpc.h [hunk 3]
```diff
@@ -74,12 +113,75 @@ CRT_RPC_DECLARE(rdbt_init, DAOS_ISEQ_RDBT_INIT_OP, DAOS_OSEQ_RDBT_INIT_OP)
...
+#define DAOS_OSEQ_RDBT_CREATE_OP	/* output fields */	 \
+	((struct rsvc_hint)	(tco_hint)		CRT_VAR) \
+	((int32_t)		(tco_rc)		CRT_VAR)
+
+CRT_RPC_DECLARE(rdbt_create, DAOS_ISEQ_RDBT_CREATE_OP,
+		DAOS_OSEQ_RDBT_CREATE_OP)
+
+#define DAOS_ISEQ_RDBT_DESTROY_OP	/* input fields (none) */
+
+#define DAOS_OSEQ_RDBT_DESTROY_OP	/* output fields */	 \
```
[the remark this answers is not in the document]
Comment (kccain) [reply src/rdb/tests/rpc.h:135]: False positive for CaRT RPC definition related macros

File: src/rdb/tests/rpc.h [hunk 4]
```diff
@@ -74,12 +113,75 @@ CRT_RPC_DECLARE(rdbt_init, DAOS_ISEQ_RDBT_INIT_OP, DAOS_OSEQ_RDBT_INIT_OP)
...
+
+#define DAOS_OSEQ_RDBT_PING_OP	/* output fields */		 \
+	((struct rsvc_hint)	(tpo_hint)		CRT_VAR) \
+	((int32_t)		(tpo_rc)		CRT_VAR)
+
+CRT_RPC_DECLARE(rdbt_ping, DAOS_ISEQ_RDBT_PING_OP, DAOS_OSEQ_RDBT_PING_OP)
+
+#define DAOS_ISEQ_RDBT_CREATE_OP	/* input fields (none) */
+
+#define DAOS_OSEQ_RDBT_CREATE_OP	/* output fields */	 \
```
[the remark this answers is not in the document]
Comment (kccain) [reply src/rdb/tests/rpc.h:126]: False positive for CaRT RPC definition related macros

File: src/rdb/tests/rpc.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (kccain) [reply src/rdb/tests/rpc.h:158]: False positive for CaRT RPC definition related macros

File: src/rdb/rdb_raft.c [hunk 5]
```diff
@@ -2521,6 +2521,18 @@ rdb_raft_resign(struct rdb *db, uint64_t term)
 	D_ASSERTF(rc == 0, ""DF_RC"\n", DP_RC(rc));
 }
 
+/* TODO: update raft.h in daos-stack/raft repo */
+extern int raft_election_start(raft_server_t *me_);
```
[the remark this answers is not in the document]
Comment (kccain) [reply src/rdb/rdb_raft.c:2525]: Will remove this and update the raft submodule once the daos-stack/raft repository PR lands https://github.com/daos-stack/raft/pull/36

File: src/rdb/tests/rdbt.c [hunk 6]
```diff
@@ -40,28 +48,70 @@ help_hdlr(int argc, char *argv[])
...
+  --replicas=N	number of replicas (1)\n\
+  --nranks=R	number of server ranks (1)\n");
 	printf("\
 test options:\n\
   --group=GROUP	server group \n\
   --rank=RANK	rank to invoke tests on (0)\n\
   --update	update (otherwise verify)\n");
+	printf("\
+fini options:\n\
+  --group=GROUP	server group \n\
```
[the remark this answers is not in the document]
Comment (kccain) [reply src/rdb/tests/rdbt.c:76]: following convention in the file, checkpatch may be getting confused about some of the string contents.

File: src/rdb/tests/rdbt.c [hunk 7]
```diff
@@ -40,28 +48,70 @@ help_hdlr(int argc, char *argv[])
...
   --group=GROUP	server group \n\
-  --rank=RANK	rank to initialize (0)\n");
+  --replicas=N	number of replicas (1)\n\
+  --nranks=R	number of server ranks (1)\n");
 	printf("\
 test options:\n\
   --group=GROUP	server group \n\
   --rank=RANK	rank to invoke tests on (0)\n\
   --update	update (otherwise verify)\n");
+	printf("\
```
[the remark this answers is not in the document]
Comment (kccain) [reply src/rdb/tests/rdbt.c:74]: following convention in the file, checkpatch may be getting confused about some of the string contents.

[the remark this answers is not in the document]
Comment (kccain) [reply src/rdb/tests/rdbt.c:74]: following convention in the file

File: src/rdb/tests/rdbt.c [hunk 8]
```diff
@@ -40,28 +48,70 @@ help_hdlr(int argc, char *argv[])
...
   --rank=RANK	rank to initialize (0)\n\
   --replicas=N	number of replicas (1)\n\
   --uuid=UUID	rdb UUID\n");
 	printf("\
-fini options:\n\
+create, test-multi, destroy options:\n\
   --group=GROUP	server group \n\
-  --rank=RANK	rank to initialize (0)\n");
+  --replicas=N	number of replicas (1)\n\
+  --nranks=R	number of server ranks (1)\n");
```
[the remark this answers is not in the document]
Comment (kccain) [reply src/rdb/tests/rdbt.c:68]: following convention in the file, checkpatch may be getting confused about some of the string contents.

File: src/rdb/tests/rdbt.c [hunk 9]
```diff
@@ -40,28 +48,70 @@ help_hdlr(int argc, char *argv[])
...
   --group=GROUP	server group \n\
   --rank=RANK	rank to initialize (0)\n\
   --replicas=N	number of replicas (1)\n\
   --uuid=UUID	rdb UUID\n");
 	printf("\
-fini options:\n\
+create, test-multi, destroy options:\n\
   --group=GROUP	server group \n\
-  --rank=RANK	rank to initialize (0)\n");
+  --replicas=N	number of replicas (1)\n\
```
[the remark this answers is not in the document]
Comment (kccain) [reply src/rdb/tests/rdbt.c:67]: following convention in the file, checkpatch may be getting confused about some of the string contents.

File: src/rdb/tests/rdbt.c [hunk 10]
```diff
@@ -40,28 +48,70 @@ help_hdlr(int argc, char *argv[])
...
+  --nranks=R	number of server ranks (1)\n");
 	printf("\
 test options:\n\
   --group=GROUP	server group \n\
   --rank=RANK	rank to invoke tests on (0)\n\
   --update	update (otherwise verify)\n");
+	printf("\
+fini options:\n\
+  --group=GROUP	server group \n\
+  --rank=RANK	rank to finalize (0)\n");
```
[the remark this answers is not in the document]
Comment (kccain) [reply src/rdb/tests/rdbt.c:77]: following convention in the file, checkpatch may be getting confused about some of the string contents.

File: src/rdb/rdb_raft.c [hunk 11]
```diff
@@ -2521,6 +2521,18 @@ rdb_raft_resign(struct rdb *db, uint64_t term)
...
+
+/* Call new election (campaign to be leader) */
+int
+rdb_raft_campaign(struct rdb *db)
+{
+	D_DEBUG(DB_MD, DF_DB": calling election from current term %d\n",
+		DP_DB(db), raft_get_current_term(db->d_raft));
+	return raft_election_start(db->d_raft);
```
Comment (liw) [inline src/rdb/rdb_raft.c:2533]: As you suggested in a mail, we need to first check

    if (i_am_not_follower)
        return -DER_NOTFOLLOWER;

Additionally, we might want to call rdb_raft_save_state and rdb_raft_check_state before and after calling raft_election_start, for consistency reasons (and perhaps error handling?).

File: src/rdb/tests/README [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+rdbt create --group=daos_server -replicas=<N> --ranks=<S>
+
+# run multi-replica tests
+rdbt test-multi --group=daos_server --replicas=<N> --nranks=<S>
+
+# destroy the KV stores
+rdbt destroy --group=daos_server -replicas=<N> --ranks=<S>
+
+# finalize the replicated database
+rdbt fini --group=daos-server
```
Comment (liw) [inline src/rdb/tests/README:36]: Thanks. Good README.

File: src/rdb/rdb_raft.c (hunk 11 above)
Comment (kccain) [reply src/rdb/rdb_raft.c]: Latest version of patch checks for not follower. As discussed in email it returns 0 if the caller is not a follower. This is similar to rdb_resign() that checks for leader status and turns that into a no-op if the condition is not met.

Comment (liw) [reply src/rdb/rdb_raft.c]: OK.

File: src/rdb/tests/README [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+Manual execution instructions:
+
+From the command line the tests are run with:
+
+==== server: start on S server ranks:
+orterun -np <S> --hostfile <file_containing_server_hostnames> --map-by-node
+ --enable-recovery -x LD_LIBRARY_PATH daos_server
+ -o <builddir>/utils/config/examples/daos_server_unittests.yml
+ start -d ./ -t 1 -m vos,rdb,rsvc,mgmt,rdbt
+
```
Comment (liw) [inline src/rdb/tests/README:14]: Sorry, forgot to ask if daos_agent is required. We do need it for attaching to work, right?

Comment (kccain) [reply src/rdb/tests/README:14]: Yes daos_agent is needed. Perhaps I could land this PR first, and include the README update to include the daos_agent item in the next DAOS PR. Another DAOS PR is needed  to update the raft submodule after Raft PR https://github.com/daos-stack/raft/pull/36 lands that will expose the raft_electon_start() function. The DAOS PR will need to remove the extern declaration of raft_election_start() from src/rdb/rdb_raft.c

Or, I could just wait for raft PR 36 to land, and update this patch with the necessary DAOS changes including the README. What do you think?

Comment (liw) [reply src/rdb/tests/README:14]: Either way is fine with me.

**Comment by kccain** [comment]
Build #7 single test failure with functional hw medium daos core test rebuild test_r_26 timeout appears to be DAOS-4626.

File: src/rdb/tests/README (hunk 13 above)
Comment (kccain) [reply src/rdb/tests/README:14]: Raft PR 36 is ready for landing, so I'll let that happen first and do the remaining steps in this PR for a single landing.

File: src/rdb/rdb_raft.c [hunk 14]
```diff
@@ -2521,6 +2521,30 @@ rdb_raft_resign(struct rdb *db, uint64_t term)
 	D_ASSERTF(rc == 0, ""DF_RC"\n", DP_RC(rc));
 }
 
+/* TODO: update raft.h in daos-stack/raft repo */
+extern int raft_election_start(raft_server_t *me_);
```
[the remark this answers is not in the document]
Comment (kccain) [reply src/rdb/rdb_raft.c]: resolved now that raft PR has landed

**Comment by kccain** [comment]
Build 8 failures traige:
1) degraded mode test failure https://jira.hpdd.intel.com/browse/DAOS-4537
2) iorsmall timeout https://jira.hpdd.intel.com/browse/DAOS-4919
3) IOR/mdtest dfuse launch failure https://jira.hpdd.intel.com/browse/DAOS-4919

For issue #3 it looks like there was another PR #2722 landed after my merge with master. Perhaps can merge with master again and run to get a slightly cleaner result...

commit f091fcdbb5cb17b959ec0cffa42fd3e75b62fbb4
Author: Phillip Henderson <phillip.henderson@intel.com>
Date: May 19, 2020 10:08 AM EDT

DAOS-4850 tests: Fixing IOR/MDTest fails with dfuse problem (#2722)

Resolving the missing environment variables needed for the dfuse
command.

Signed-off-by: Phillip Henderson <phillip.henderson@intel.com>

**Comment by kccain** [comment]
It appears that merging with latest master to try to get a better mdtest outcome may not work. There is some ongoing discussion on IM channels about this. After some discussion with Li Wei, adding gatekeeper now for review/landing.

**Comment by kccain** [comment]
> Looks like you lost a race on daos.spec that need to be resolved

Merged that daos.spec and fixed the conflict. Have pushed another version of the patch. Will make sure it cleanly gets through the build steps and then request your review.

**Pull request merged**